### DISCUSSION TITLE
Declaring return types in method definition restricted to the declared method

### DISCUSSION THREAD

**Discussion by mcurros2** [dotnet/csharplang#6840] [Language Ideas]
I am searching a way to solve the problem of _**defining a specific class/record/struct for a method return type that should be restricted only as a return type for that specific method**_.

Pseudocode for the best I can do with tuples:

```CSharp
       // You can't rename tuple fields with visual studio
       // if you change the name of one field, I receive a warning that the tuple fieldname doesn't match
       // Is not possible to use nameof() with a tuple field name
        public (
            string emaildestination_id
            , string email
            , string name
            , string messagegroup_id)
            GetNextPendingEmail()
        {
                // Code for fetching data here
               ....

               // I can create the tuple only on return
               // yields intellisense with the declared field names
               // BAD: I can use named parameters but they don't work, it gives the illusion it does. May be a bug in VS?
               return (emaildestination_id: "1", email: "john@nowhere.com", name: "john doe", messagegroup_id: "test");

        }

        // It forces me to declare the tuple inside the method, I need to create a return list before processing the data
        public List<(
            string emaildestination_id
            , string email
            , string name
            , string messagegroup_id)>
            GetNextPendingEmail()
        {
                // This yields the declared return type tuple useless
               var result = new List<(
            string emaildestination_id
            , string email
            , string name
            , string messagegroup_id)>();

           ...
        }

        public async IAsyncEnumerable<(
            string emaildestination_id
            , string email
            , string name
            , string messagegroup_id)>
            GetNextPendingEmail()
        {
                // This works as a workaround for returning a list of tuples without declaring it twice
               yield return (emaildestination_id: r[1], email: r[2], name: r[3], messagegroup_id: r[0]);

        }

```

This is the version with a specific nested class to return as a result for the same method:

```CSharp
        public sealed class GetNextPendingEmailsResult
        {
            public string emaildesination_id { get; init; }
            public string email { get; init; }
            public string name { get; init; }
            public string messagegroup { get; init; }

            internal GetNextPendingEmailsResult()
            {

            }
        }

        public async Task<List<GetNextPendingEmailsResult>>
            GetNextPendingEmails(CancellationToken ct)
        {
            var result = new List<GetNextPendingEmailsResult>();
            var data = await Data.ExecuteProc(ct, Def.nque_GetNextPendingEmails);
            if (data.HasData())
            {
                var res = data[0];
                for (int x = 0; x < res.records.Count; x++)
                {
                    if (ct.IsCancellationRequested) break;
                    result.Add(new()
                    {
                        emaildesination_id = (string)res[x, nameof(GetNextPendingEmailsResult.emaildesination_id)],
                        email = (string)res[x, nameof(GetNextPendingEmailsResult.email)],
                        name = (string)res[x, nameof(GetNextPendingEmailsResult.name)],
                        messagegroup = (string)res[x, nameof(GetNextPendingEmailsResult.messagegroup)]
                    });
                }
            }
            return result;
        }

```
The above will force me to declare a class as return type for every method like the above. I understand that there will be situations that need to return more complex data that will require a specific class, so I will stick with basic things.

It would be a useful feature if C# has the ability to infer the returning type at compile time. Something like:

```CSharp
     public [var] GetEmails() // I couldn't come up with a keyword to indicate the return type is inferred
     {
          List<(string, name, string email)> result = new();

         ....

          return result;
     }

```
Or be able to declare a var of the return type defined ad hoc for the method

```CSharp
     public List<(string, name, string email)> GetEmails()
     {
           var result = [returntype] new(); // returntype would be a new keyword

         ....

          return result;
     }

```
Instead of declaring tuples I would like to use something that can infer a class like declaration, so it can support renaming and nameof(). Something like:

```CSharp
     public List<{string name, string email}> GetEmails() // Note the { } that will indicate the treatment as a class and not a Tuple
     {
           var result = [returntype] new(); // returntype would be a new keyword

         ....

          return result;
     }

```
Do you think that it can be something viable?

Cheers!

**Comment by iam3yal** [comment]
@mcurros2 I'm not completely sure but I think that the compiler doesn't look inside the method body to get information about the signature of a method and at the same time I think that in these cases you better use a dedicated type anyway so it's not worth the added complexity, also, you better break these requests into 2 different requests because the first is about inferring tuples return types from the method body whereas the other is not really related to type inference but about some unexplained chemistry between the syntax `{}` and the keyword `returntype` that by the looks of it both of your issues can be addressed or mitigated with something like type aliases and there are existing championed proposal and discussions about it on multiple fronts.

**Reply by mcurros2** [reply]
I just was trying not to declare a class for the return type of each method. I haven't explained too well the problem. I'm working on an ERP that has 10K+ different queries that return different results. Those results are not entities, are specific data to accomplish different tasks. Entities and ORM mapping are not the problem.

Most ORMs return dynamic objects for these cases but is difficult to make renames and intellisense doesn't work. So, I don't have a problem with defining a class or specific type for each result. I was just wondering if I can do it when declaring the method that gets the data. That's when I used tuples with some success, but tuples field names get lost, can't be renamed.

If tuples preserve the field names, would be a viable choice. The weird syntax with { } was a poor intent to somehow declare a sort of anonymous type.

I have seen many other discussions about tuple item names, but it seems that it won't never happen.

Anyways, thanks for the reply.

**Comment by svick** [comment]
I think using a `record` does help a lot here:

```c#
public record GetNextPendingEmailResult(
    string emaildestination_id,
    string email,
    string name,
    string messagegroup_id);
```

**Reply by mcurros2** [reply]
Thanks! It would be nice to have something like:

```CSharp
        public readonly record struct (string email, string name) GetEmail()
        {

        }
```
It looks better than tuples, but I think will carry the same limitations.

**Reply by svick** [reply]
But you also complained that `new List<ResultType>()` is too verbose with a tuple. A regular named record type also solves that nicely, the suggested anonymous record type wouldn't.

Also, this would help with your tooling issue with tuples, but the language is much harder to change than tooling, so if something can be done by improving tooling, it should, instead of changing the language.

**Reply by mcurros2** [reply]
Sorry if it sounded I was complaining. Just thinking about it. I guess I was just looking a way to inline the return type declaration for a method. It seems it adds complexity with little value.

Cheers.

**Reply by CyrusNajmabadi** [reply]
The main issue here is that say you do the above. i.e.: 

```c#
        public readonly record struct (string email, string name) GetEmail()
        {

        }
```

Now you want to write:

```c#
ProcessEmail(GetEmail());
```

How do you write teh signature for `ProcessEmail`?  A record allows you to define this type (i.e. give a name to it) with minimal effort and very clear code.  You can then reference that record anywhere you need it (like as a parameter to `ProcessEmail`).  

If you don't do that, you have no way to reference it, and would have to spell out the entire structure again.  But, as you mentioned, that would be excessively verbose.  It seems like you don't want either approach, but i'm not seeing a viable 3rd option that solve this from you.  thanks!

**Reply by mcurros2** [reply]
You are right. 

When I was thinking about this I thought about Anonymous Types. The compiler does some magic with anonymous types: it emits a sealed class that has assembly scope with a complex name and keep it hidden. I was thinking about the compiler generating the class with a name that can tie it to the method with the inlined declaration,  something like $"{class_name}_{method_name}_ResultType".

For the sample given:

```CSharp

public readonly record struct (string email, string name) GetEmail();

// If you would like to declare a property of the type on another class
ContainingClass.GetEmail.ResultType email;

ProcessEmail(GetEmail()); // should work as the inlined return type is "visible" to the compiler.

```

Anyways it seems too complex to implement and will add little value to specific use cases when the return type is a "simple type" like a record.

Thanks!

**Reply by CyrusNajmabadi** [reply]
You could do this with a source-generator.  just write the method as `public MyClassName_GetEmail_ResultType GetEmail() { ... }`  then your generator will generate the actual record (or class or struct or whatever) for that based on what goes on in `GetEmail`.  I'm not sure if this is really taht valuable though as writing the return-type manually is just a single simple line.
